From d57c72e45da4b24948232701a953ea65dd1cd1b7 Mon Sep 17 00:00:00 2001
From: Jonas Karlman <jonas@kwiboo.se>
Date: Fri, 24 Apr 2026 07:50:10 +0000
Subject: [PATCH 164/280] WIP: drm/bridge: dw-hdmi: filter input bus formats by
 supported_formats

dw-hdmi_bridge_atomic_get_input_bus_fmts() currently returns candidate
input bus formats based only on the requested output format. This can
advertise RGB, YUV444, YUV422 or YUV420 input formats that the upstream
encoder cannot actually produce.

Filter the returned input format list against plat_data->supported_formats
so only encoder-supported formats are exposed during bridge format
negotiation.

Keep the existing default behavior by falling back to RGB444 when
supported_formats is not set.

Signed-off-by: Jonas Karlman <jonas@kwiboo.se>
---
 drivers/gpu/drm/bridge/synopsys/dw-hdmi.c | 104 +++++++++++++++-------
 1 file changed, 72 insertions(+), 32 deletions(-)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
index a3be17d48729..fc783a2225f3 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi.c
@@ -2736,6 +2736,8 @@ static u32 *dw_hdmi_bridge_atomic_get_input_bus_fmts(struct drm_bridge *bridge,
 					u32 output_fmt,
 					unsigned int *num_input_fmts)
 {
+	struct dw_hdmi *hdmi = bridge->driver_private;
+	unsigned int supported_formats;
 	u32 *input_fmts;
 	unsigned int i = 0;
 
@@ -2746,6 +2748,12 @@ static u32 *dw_hdmi_bridge_atomic_get_input_bus_fmts(struct drm_bridge *bridge,
 	if (!input_fmts)
 		return NULL;
 
+	supported_formats = hdmi->plat_data->supported_formats;
+
+	/* Default format expected from encoder is full range RGB */
+	if (!supported_formats)
+		supported_formats = BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444);
+
 	switch (output_fmt) {
 	/* If MEDIA_BUS_FMT_FIXED is tested, return default bus format */
 	case MEDIA_BUS_FMT_FIXED:
@@ -2753,63 +2761,94 @@ static u32 *dw_hdmi_bridge_atomic_get_input_bus_fmts(struct drm_bridge *bridge,
 		break;
 	/* 8bit */
 	case MEDIA_BUS_FMT_RGB888_1X24:
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
 		break;
 	case MEDIA_BUS_FMT_YUV8_1X24:
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
 		break;
 	case MEDIA_BUS_FMT_UYVY8_1X16:
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY8_1X16;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV8_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB888_1X24;
 		break;
 
 	/* 10bit */
 	case MEDIA_BUS_FMT_RGB101010_1X30:
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
 		break;
 	case MEDIA_BUS_FMT_YUV10_1X30:
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
 		break;
 	case MEDIA_BUS_FMT_UYVY10_1X20:
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY10_1X20;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV10_1X30;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB101010_1X30;
 		break;
 
 	/* 12bit */
 	case MEDIA_BUS_FMT_RGB121212_1X36:
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
 		break;
 	case MEDIA_BUS_FMT_YUV12_1X36:
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
 		break;
 	case MEDIA_BUS_FMT_UYVY12_1X24:
-		input_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR422))
+			input_fmts[i++] = MEDIA_BUS_FMT_UYVY12_1X24;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV12_1X36;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB121212_1X36;
 		break;
 
 	/* 16bit */
 	case MEDIA_BUS_FMT_RGB161616_1X48:
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB161616_1X48;
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV16_1X48;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB161616_1X48;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV16_1X48;
 		break;
 	case MEDIA_BUS_FMT_YUV16_1X48:
-		input_fmts[i++] = MEDIA_BUS_FMT_YUV16_1X48;
-		input_fmts[i++] = MEDIA_BUS_FMT_RGB161616_1X48;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR444))
+			input_fmts[i++] = MEDIA_BUS_FMT_YUV16_1X48;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_RGB444))
+			input_fmts[i++] = MEDIA_BUS_FMT_RGB161616_1X48;
 		break;
 
 	/*YUV 4:2:0 */
@@ -2817,7 +2856,8 @@ static u32 *dw_hdmi_bridge_atomic_get_input_bus_fmts(struct drm_bridge *bridge,
 	case MEDIA_BUS_FMT_UYYVYY10_0_5X30:
 	case MEDIA_BUS_FMT_UYYVYY12_0_5X36:
 	case MEDIA_BUS_FMT_UYYVYY16_0_5X48:
-		input_fmts[i++] = output_fmt;
+		if (supported_formats & BIT(DRM_OUTPUT_COLOR_FORMAT_YCBCR420))
+			input_fmts[i++] = output_fmt;
 		break;
 	}
 
-- 
2.43.0

